test_report/src/report.rs

```diff
@@ -0,0 +1,268 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Status {
+    Success,
+    Failure,
+    Skipped,
+}
+
+#[cfg(feature = "ruby")]
+impl Status {
+    fn new(status: String) -> Self {
+        match status.as_str() {
+            "success" => Status::Success,
+            "failure" => Status::Failure,
+            "skipped" => Status::Skipped,
+            _ => Status::Success,
+        }
+    }
+}
+
+impl Into<&str> for Status {
+    fn into(self) -> &'static str {
+        match self {
+            Status::Success => "success",
+            Status::Failure => "failure",
+            Status::Skipped => "skipped",
+        }
+    }
+}
+
+impl ToString for Status {
+    fn to_string(&self) -> String {
+        String::from(Into::<&str>::into(*self))
+    }
+}
+
+#[cfg(feature = "ruby")]
+impl Status {
+    pub fn to_string(&self) -> &str {
+        (*self).into()
+    }
+}
+
+#[cfg(feature = "ruby")]
+impl magnus::TryConvert for Status {
+    fn try_convert(val: magnus::Value) -> Result<Self, magnus::Error> {
+        let sval: String = val.funcall("to_s", ())?;
+        Ok(Status::new(sval))
+    }
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen(getter_with_clone))]
+#[cfg_attr(feature = "ruby", magnus::wrap(class = "TestReport"))]
+#[derive(Debug, Clone, PartialEq)]
+pub struct MutTestReport(RefCell<TestReport>);
+
+#[cfg_attr(feature = "wasm", wasm_bindgen)]
+impl MutTestReport {
+    pub fn new(origin: String) -> Self {
+        let mut test_result = TestResult::default();
+        test_result.uploader_metadata = Some(UploaderMetadata {
+            origin,
+            version: std::env!("CARGO_PKG_VERSION").to_string(),
+            upload_time: None,
+        });
+        Self(RefCell::new(TestReport {
+            test_result: TestResult::default(),
+        }))
+    }
+
+    fn serialize_test_result(&self) -> Vec<u8> {
+        let test_result = self.0.borrow().test_result.clone();
+        let buf: Vec<u8> = prost::Message::encode_to_vec(&test_result);
+        buf
+    }
+
+    // sends out to the trunk api
+    pub fn publish(&self) -> bool {
+        let path = self.save();
+        if path.is_err() {
+            return false;
+        }
+        let resolved_path = path.unwrap_or_default();
+        let token = std::env::var("TRUNK_API_TOKEN").unwrap_or_default();
+        let org_url_slug = std::env::var("TRUNK_ORG_URL_SLUG").unwrap_or_default();
+        if token.is_empty() || org_url_slug.is_empty() {
+            println!("Token or org url slug not set");
+            return false;
+        }
+        if let Some(uploader_metadata) = &mut self.0.borrow_mut().test_result.uploader_metadata {
+            uploader_metadata.upload_time = Some(Timestamp {
+                seconds: Utc::now().timestamp(),
+                nanos: Utc::now().timestamp_subsec_nanos() as i32,
+            });
+        }
+        let upload_args = trunk_analytics_cli::upload::UploadArgs::new(
```

Comment:
This is where the magic happens. We invoke all of the same upload logic as the cli to reduce fragmentation.

Reply:
Am I missing something? Don't we need the upload.rs logic to be able to handle `.bin` files somewhere?